File: packages/pvm/src/host-functions/general/gas.ts
import { logger } from '@pbnjam/core'
import type { HostFunctionContext, HostFunctionResult } from '@pbnjam/types'
import { GENERAL_FUNCTIONS } from '../../config'
import { BaseHostFunction } from './base'

/**
 * GAS host function (Ω_G)
 *
 * Returns the current gas counter value
 *
 * Gray Paper Specification:
 * - Function ID: 0 (gas)
 * - Gas Cost: 10
 * - Sets registers[7] = gascounter (remaining gas)
 * - Returns gascounter - 10 (gas consumed)
 */
export class GasHostFunction extends BaseHostFunction {
  readonly functionId = GENERAL_FUNCTIONS.GAS
  readonly name = 'gas'

  execute(context: HostFunctionContext): HostFunctionResult {
    context.registers[7] = context.gasCounter

    // Log in the requested format: [host-calls] [serviceId] GAS <- gasValue
    const serviceId = context.serviceId ?? 0n
    logger.info(`[host-calls] [${serviceId}] GAS <- ${context.gasCounter}`)

    // Return updated gas counter
    return {
      resultCode: null, // continue execution
    }
  }
}

File: packages/pvm-rust/src/host_functions/general/gas.rs
//! GAS host function (Ω_G). Gray Paper: function ID 0, gas cost 10.
//! Sets registers[7] = remaining gas.

use crate::config::FUNC_GAS;
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

/// GAS (0): set registers[7] = remaining gas. Base 10 gas already deducted in state_wrapper.
pub struct GasHostFunction;

impl HostFunction for GasHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_GAS
    }
    fn name(&self) -> &'static str {
        "gas"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        context.registers[7] = u64::from(*context.gas_remaining);
        HostFunctionResult::continue_execution()
    }
}
